compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs

```diff
@@ -21,104 +21,224 @@ use crate::hir_ty_lowering::{
+        for hir_bound in hir_bounds {
+            let hir::GenericBound::Trait(ptr) = hir_bound else {
+                continue;
+            };
+
+            if matches!(ptr.modifiers.polarity, hir::BoundPolarity::Maybe(_)) {
+                unbounds.push(ptr);
+            }
+
+            if ptr.trait_ref.path.res == Res::Def(DefKind::Trait, sized_did) {
                 match ptr.modifiers.polarity {
-                    hir::BoundPolarity::Maybe(_) => unbounds.push(ptr),
-                    hir::BoundPolarity::Negative(_) => {
-                        if let Some(sized_def_id) = sized_def_id
-                            && ptr.trait_ref.path.res == Res::Def(DefKind::Trait, sized_def_id)
-                        {
-                            seen_negative_sized_bound = true;
-                        }
-                    }
-                    hir::BoundPolarity::Positive => {
-                        if let Some(sized_def_id) = sized_def_id
-                            && ptr.trait_ref.path.res == Res::Def(DefKind::Trait, sized_def_id)
-                        {
-                            seen_positive_sized_bound = true;
-                        }
-                    }
+                    hir::BoundPolarity::Maybe(_) => sized.maybe = true,
+                    hir::BoundPolarity::Negative(_) => sized.negative = true,
+                    hir::BoundPolarity::Positive => sized.positive = true,
                 }
             }
-        };
-        search_bounds(hir_bounds);
-        if let Some((self_ty, where_clause)) = self_ty_where_predicates {
-            for clause in where_clause {
-                if let hir::WherePredicateKind::BoundPredicate(pred) = clause.kind
-                    && pred.is_param_bound(self_ty.to_def_id())
-                {
-                    search_bounds(pred.bounds);
+
+            if ptr.trait_ref.path.res == Res::Def(DefKind::Trait, metasized_did) {
+                match ptr.modifiers.polarity {
+                    hir::BoundPolarity::Maybe(_) => metasized.maybe = true,
+                    hir::BoundPolarity::Negative(_) => metasized.negative = true,
+                    hir::BoundPolarity::Positive => metasized.positive = true,
                 }
             }
-        }
 
-        let mut unique_bounds = FxIndexSet::default();
-        let mut seen_repeat = false;
-        for unbound in &unbounds {
-            if let Res::Def(DefKind::Trait, unbound_def_id) = unbound.trait_ref.path.res {
-                seen_repeat |= !unique_bounds.insert(unbound_def_id);
+            if ptr.trait_ref.path.res == Res::Def(DefKind::Trait, pointeesized_did) {
+                match ptr.modifiers.polarity {
+                    hir::BoundPolarity::Maybe(_) => pointeesized.maybe = true,
+                    hir::BoundPolarity::Negative(_) => pointeesized.negative = true,
+                    hir::BoundPolarity::Positive => pointeesized.positive = true,
+                }
             }
```

Comment:
This part can be deduplicated through doing something like

```rust
let collected = if ... { &mut sized } else if ... { &mut meta_sized } else if { &mut pointee_sized } else { continue };  
```

Reply:
Fixed